### ISSUE TITLE
bad result

### ISSUE THREAD

**Issue by melih1996** [chaitanya100100/TailorNet#18]
Hi, when I put pose parameters from another model, I get bad results for the right side of the body. How to resolve it?
![Screenshot from 2021-02-01 22-00-58](https://user-images.githubusercontent.com/34304254/106505453-63348800-64d9-11eb-97f4-5f9d93eba726.png)
![Screenshot from 2021-02-01 22-01-49](https://user-images.githubusercontent.com/34304254/106505458-6465b500-64d9-11eb-82e4-bdc016191872.png)
![Screenshot from 2021-02-01 22-02-59](https://user-images.githubusercontent.com/34304254/106505463-6596e200-64d9-11eb-97f3-3fbd279e0624.png)

**Comment by melih1996** [comment]
![Uploading Screenshot from 2021-02-01 22-06-24.png…]()

**Comment by melih1996** [comment]
![Screenshot from 2021-02-01 22-06-24](https://user-images.githubusercontent.com/34304254/106505801-ddfda300-64d9-11eb-950a-f085d76f1ddc.png)

**Comment by melih1996** [comment]
I have also faced this kind of issue in the skirt:
![Screenshot from 2021-02-02 11-45-17](https://user-images.githubusercontent.com/34304254/106574695-47b69500-654c-11eb-9207-38da4102f175.png)

**Comment by chaitanya100100** [comment]
We can't figure out the issue just from images. Can you provide all the necessary details like theta, beta and gamma parameters to reproduce at least one of these results?

**Comment by melih1996** [comment]
[params.zip](https://github.com/chaitanya100100/TailorNet/files/5910412/params.zip)
def get_single_frame_inputs(garment_class, gender,melih=False):
    """Prepare some individual frame inputs."""
    betas = [
        np.load("my_betas.npy")[0],
    ]
    
    # old t-shirt style parameters are centered around [1.5, 0.5, 1.5, 0.0]
    # whereas all other garments styles are centered around [0, 0, 0, 0]
    if garment_class == 'old-t-shirt':
        gammas = [
            get_specific_style_old_tshirt('big'),
        ]
    else:
        gammas = [
            get_style('008', garment_class=garment_class, gender=gender),
        ]
    thetas = [
        np.load("my_body_pose.npy")[0],
    ]
    return thetas, betas, gammas

**Comment by chaitanya100100** [comment]
It would also help if you can provide little code snippet to reproduce one of these results.

**Comment by chaitanya100100** [comment]
In other words, please provide a short code where you hard-code theta, beta and gamma; load models; pass inputs and get the output. A short code which we can run directly to reproduce these results.

**Comment by melih1996** [comment]
`
def get_single_frame_inputs(garment_class, gender):
    """Prepare some individual frame inputs."""
    betas = [
        np.array([ 0.02372807,  0.2726095 ,  0.2081716 ,  0.38198256,  0.15542014,
         0.06538279, -0.38254178,  0.28198612,  0.69489276,  0.12113538])
    ]

    if garment_class == 'old-t-shirt':
        gammas = [
            get_specific_style_old_tshirt('big'),
        ]
    else:
        gammas = [
            np.array([-0.5,  0. ,  0. ,  0. ]) if garment_class == "shirt" else np.array([-0.33333334,  0., 0., 0.])
        ]
    thetas = [
        np.array([-3.02273726e+00,  3.74335736e-01, -6.63577840e-02,
        -3.71137649e-01,  4.41651903e-02, -7.39246756e-02,
        -3.92500907e-01,  3.12881656e-02, -2.18228787e-01,
         2.93597966e-01,  2.47394812e-04,  8.06266740e-02,
         6.35661662e-01, -5.46876192e-02, -9.93532985e-02,
         6.68145299e-01,  1.87627152e-01,  3.60645503e-02,
        -9.25337449e-02,  2.95514669e-02,  1.61951452e-01,
        -7.18512535e-02,  2.76237607e-01, -1.46155521e-01,
         2.45762169e-02, -2.39159718e-01,  2.44792506e-01,
         3.79879847e-02,  4.06366661e-02,  1.04108125e-01,
        -1.36675581e-01, -2.87497416e-02,  2.91092932e-01,
        -2.22747564e-01,  1.13584250e-01, -3.86912435e-01,
        -2.68053144e-01, -1.02593645e-01,  3.91437151e-02,
        -2.29289368e-01,  5.09471931e-02,  2.25034893e-01,
        -5.24876714e-02,  8.59650448e-02,  2.21216157e-01,
        -1.84436321e-01, -2.45631918e-01,  5.59661873e-02,
        -4.64416504e-01, -1.92874238e-01,  9.88448709e-02,
         1.21609941e-02,  3.92712280e-03,  6.50697172e-01,
        -5.35148203e-01, -8.60135436e-01,  4.68042523e-01,
        -1.71674415e-01,  3.32635045e-01, -1.14242852e-01,
         6.91787377e-02, -1.39364123e-01,  1.43555686e-01,
        -6.67377189e-02,  7.22975433e-02, -4.70257215e-02,
        -8.51052254e-02, -5.07538058e-02, -1.54520690e-01,
        -5.57201654e-02,  6.73477277e-02,  1.20280601e-01])
    ]
    return thetas, betas, gammas

def get_input(garment_class, gender):
    smpl = SMPL4Garment(gender=gender)
    thetas, betas, gammas = get_single_frame_inputs(garment_class, gender)
    tn_runner = get_tn_runner(gender=gender, garment_class=garment_class)
    theta=thetas[0]
    beta=betas[0]
    gamma=gammas[0]
    theta_normalized = normalize_y_rotation(theta)
    with torch.no_grad():
        pred_verts_d = tn_runner.forward(
            thetas=torch.from_numpy(theta_normalized[None, :].astype(np.float32)).cuda(),
            betas=torch.from_numpy(beta[None, :].astype(np.float32)).cuda(),
            gammas=torch.from_numpy(gamma[None, :].astype(np.float32)).cuda(),
            )[0].cpu().numpy()
    body, pred_gar = smpl.run(beta=beta, theta=theta, garment_class=garment_class, garment_d=pred_verts_d)
    pred_gar = remove_interpenetration_fast(pred_gar, body)
    return body,pred_gar

def run_and_render_two_garments():
    gender="female"
    garment_up_class = 'shirt'
    garment_down_class = 'skirt'
    _,upper=get_input(garment_up_class,gender)
    body,lower=get_input(garment_down_class,gender)
    visualize_two_garments_body(lower, upper, body, f"{OUT_PATH}/out.png", lower_gc=garment_down_class, upper_gc=garment_up_class, side='front')`

**Comment by melih1996** [comment]
Also intrestingly the the results for pants are not good:
![Screenshot from 2021-02-02 14-47-36](https://user-images.githubusercontent.com/34304254/106596110-9e7c9880-6565-11eb-82b4-b45105b5f234.png)

**Comment by zycliao** [comment]
> ```
> def get_single_frame_inputs(garment_class, gender):
> """Prepare some individual frame inputs."""
> betas = [
> np.array([ 0.02372807, 0.2726095 , 0.2081716 , 0.38198256, 0.15542014,
> 0.06538279, -0.38254178, 0.28198612, 0.69489276, 0.12113538])
> ]
> gammas = [
> np.array([-0.5, 0. , 0. , 0. ]) if garment_class == "shirt" else np.array([-0.33333334, 0., 0., 0.])
> ]
> thetas = [
> np.array([-3.02273726e+00, 3.74335736e-01, -6.63577840e-02,
> -3.71137649e-01, 4.41651903e-02, -7.39246756e-02,
> -3.92500907e-01, 3.12881656e-02, -2.18228787e-01,
> 2.93597966e-01, 2.47394812e-04, 8.06266740e-02,
> 6.35661662e-01, -5.46876192e-02, -9.93532985e-02,
> 6.68145299e-01, 1.87627152e-01, 3.60645503e-02,
> -9.25337449e-02, 2.95514669e-02, 1.61951452e-01,
> -7.18512535e-02, 2.76237607e-01, -1.46155521e-01,
> 2.45762169e-02, -2.39159718e-01, 2.44792506e-01,
> 3.79879847e-02, 4.06366661e-02, 1.04108125e-01,
> -1.36675581e-01, -2.87497416e-02, 2.91092932e-01,
> -2.22747564e-01, 1.13584250e-01, -3.86912435e-01,
> -2.68053144e-01, -1.02593645e-01, 3.91437151e-02,
> -2.29289368e-01, 5.09471931e-02, 2.25034893e-01,
> -5.24876714e-02, 8.59650448e-02, 2.21216157e-01,
> -1.84436321e-01, -2.45631918e-01, 5.59661873e-02,
> -4.64416504e-01, -1.92874238e-01, 9.88448709e-02,
> 1.21609941e-02, 3.92712280e-03, 6.50697172e-01,
> -5.35148203e-01, -8.60135436e-01, 4.68042523e-01,
> -1.71674415e-01, 3.32635045e-01, -1.14242852e-01,
> 6.91787377e-02, -1.39364123e-01, 1.43555686e-01,
> -6.67377189e-02, 7.22975433e-02, -4.70257215e-02,
> -8.51052254e-02, -5.07538058e-02, -1.54520690e-01,
> -5.57201654e-02, 6.73477277e-02, 1.20280601e-01])
> ]
> return thetas, betas, gammas
> 
> def get_input(garment_class, gender):
> smpl = SMPL4Garment(gender=gender)
> thetas, betas, gammas = get_single_frame_inputs(garment_class, gender)
> tn_runner = get_tn_runner(gender=gender, garment_class=garment_class)
> theta=thetas[0]
> beta=betas[0]
> gamma=gammas[0]
> theta_normalized = normalize_y_rotation(theta)
> with torch.no_grad():
> pred_verts_d = tn_runner.forward(
> thetas=torch.from_numpy(theta_normalized[None, :].astype(np.float32)).cuda(),
> betas=torch.from_numpy(beta[None, :].astype(np.float32)).cuda(),
> gammas=torch.from_numpy(gamma[None, :].astype(np.float32)).cuda(),
> )[0].cpu().numpy()
> body, pred_gar = smpl.run(beta=beta, theta=theta, garment_class=garment_class, garment_d=pred_verts_d)
> pred_gar = remove_interpenetration_fast(pred_gar, body)
> return body,pred_gar
> 
> def run_and_render_two_garments():
> gender="female"
> garment_up_class = 'shirt'
> garment_down_class = 'skirt'
> _,upper=get_input(garment_up_class,gender)
> body,lower=get_input(garment_down_class,gender)
> visualize_two_garments_body(lower, upper, body, f"{OUT_PATH}/out.png", lower_gc=garment_down_class, upper_gc=garment_up_class, side='front')
> 
> ```

Hi, I notice the theta you use make the upward direction of the human body be -y. You should make sure the up direction is +y. For example, if you make it rotate around x for 180 degree (the following code is an example), the result will be normal.
```
thetas = [np.array([... , ...])]
glob_rot = thetas[0][:3]
r = cv2.Rodrigues(glob_rot)[0]
r = np.array([[1, 0, 0], [0, -1, 0], [0, 0, -1]]) @ r
glob_rot = cv2.Rodrigues(r)[0][:, 0]
thetas[0][:3] = glob_rot
```
![img_0000](https://user-images.githubusercontent.com/25116439/106632948-954bf580-657e-11eb-81a3-09a01aa8d397.png)

**Comment by HugoBA92** [comment]
hi, @melih1996 

if you are having issues with TailorNet, you might be interested in this:
https://github.com/HugoBA92/PBNS

it is a surprisingly simple approach to garment animation with strong results,
it works unsupervisedly and training takes barely a few minutes,
additionally, it is compatible with any animated 3D model (not just SMPL),

in the git you will find a link to a 17min long video of qualitative results (with NO post-processing),

I hope this is useful! ^^

**Comment by melih1996** [comment]
@HugoBA92 which dataset to train the model? I also need fast inference so what is the inference time of a single image

**Comment by HugoBA92** [comment]
Hi @melih1996 ,

To train this model you only need pose data.
For our experiments, we use AMASS dataset, and more specifically, CMU.
It is important that training poses are balanced.
I recommend to you reading Sec.4.3 of the related paper (link in the git).

Regarding inference time, it should take around 6ms to process a single sample (outfit + body). You can speed this up by processing multiple samples at a time (if your applications permits this).
Additionally, further pruning experiments show that our model can be reduced to a 1/8 of its size and produce the same results (therefore reducing computational cost to a 1/8).

Note that our model just generates 3D data, rendering should be performed with another software (I highly recommend Blender).
As it is now, the script 'eval.py' computes animation data for a given pose matrix (T x 72). It saves results as OBJ and PC2 files, both are easily imported into Blender.
If your application does not permit pre-computing animation data, note that our methodology yields blend weights and a blend shape matrix for pose space deformations for the outfit. This should allow easy integration of the model into real-time animation pipelines.

I hope this could help you, and please do not hesitate to ask further questions =)
